zerver/tests/test_topic_visibility_policies.py

```diff
@@ -107,29 +108,33 @@ def test_add_muted_topic(self) -> None:
             self.assertIn((stream.name, "Verona3", mock_date_muted), get_topic_mutes(user))
             self.assertTrue(topic_is_muted(user, stream.id, "verona3"))
 
-            remove_topic_mute(
+            set_user_topic_visibility_policy(
                 user_profile=user,
                 stream_id=stream.id,
                 topic_name="Verona3",
+                visibility_policy=UserTopic.VISIBILITY_POLICY_INHERIT,
             )
 
-        # Verify the error handling for the database level
-        # IntegrityError we'll get with a race between two processes
-        # trying to mute the topic.  To do this, we patch the
-        # topic_is_muted function to always return False when trying
-        # to mute a topic that is already muted.
-        assert stream.recipient is not None
-        add_topic_mute(
-            user_profile=user,
-            stream_id=stream.id,
-            recipient_id=stream.recipient.id,
-            topic_name="Verona3",
-            date_muted=datetime(2020, 1, 1, tzinfo=timezone.utc),
-        )
-
-        with mock.patch("zerver.views.muting.topic_is_muted", return_value=False):
-            result = self.api_patch(user, url, data)
-            self.assert_json_error(result, "Topic already muted")
+        # DISCUSSION REQUIRED FOR THIS PIECE OF TEST-CASE
+
+        # # Verify the error handling for the database level
+        # # IntegrityError we'll get with a race between two processes
+        # # trying to mute the topic.  To do this, we patch the
+        # # topic_is_muted function to always return False when trying
+        # # to mute a topic that is already muted.
+        # assert stream.recipient is not None
+        # set_user_topic_visibility_policy(
+        #     user_profile=user,
+        #     stream_id=stream.id,
+        #     recipient_id=stream.recipient.id,
+        #     topic_name="Verona3",
+        #     visibility_policy=UserTopic.MUTED,
+        #     date_added=datetime(2020, 1, 1, tzinfo=timezone.utc),
+        # )
+
+        # with mock.patch("zerver.views.user_topics.topic_is_muted", return_value=False):
+        #     result = self.api_patch(user, url, data)
+        #     self.assert_json_error(result, "Topic already muted")
```

Comment:
This needs discussion:

In our [`set_user_topic_visibility_policy`](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR108) function,

If the {"user_profile", "stream", "topic_name"} triplet is already present in `UserTopic` table. Update the `visibility_policy`.
Otherwise, create a new entry.

But this leads to the situation where we are not handling `Integrity Error`, i.e. Already Muted.
Of course, `mute_topic` handles this using `topic_is_muted`. But as mentioned here -- race-condition -- how should we handle this?

Fetching + comparing will lead to the addition of extra queries.

Reply:
Not an answer to the OP, but I did not entirely understand how the earlier works: https://github.com/zulip/zulip/pull/21535#issuecomment-1416697323
Maybe having an answer to that question would give more context as to how the commit in this PR should be.

Reply:
Please start a #backend conversation about this; I replied to https://github.com/zulip/zulip/pull/21535#issuecomment-1416697323 with some additional context. But I think the right answer is that we should ONLY have the database-level integrity check ... but that we might need to add a missing case-insensitive index and then we'd be able to only have the `IntegrityError` check, and get coverage on it through a straightforward "I tried to upload a thing twice" test.

```
zulip=> \d zerver_usertopic;
                                     Table "zulip.zerver_usertopic"
      Column       |           Type           | Collation | Nullable |             Default              
-------------------+--------------------------+-----------+----------+----------------------------------
 id                | integer                  |           | not null | generated by default as identity
 topic_name        | character varying(60)    |           | not null | 
 recipient_id      | integer                  |           | not null | 
 stream_id         | integer                  |           | not null | 
 user_profile_id   | integer                  |           | not null | 
 last_updated      | timestamp with time zone |           | not null | 
 visibility_policy | smallint                 |           | not null | 
Indexes:
    "zerver_mutedtopic_pkey" PRIMARY KEY, btree (id)
    "zerver_mutedtopic_recipient_id_e1901302" btree (recipient_id)
    "zerver_mutedtopic_stream_id_acbff20e" btree (stream_id)
    "zerver_mutedtopic_stream_topic" btree (stream_id, upper(topic_name::text))
    "zerver_mutedtopic_user_profile_id_4f8a692c" btree (user_profile_id)
    "zerver_mutedtopic_user_profile_id_stream_i_2cb30f72_uniq" UNIQUE CONSTRAINT, btree (user_profile_id, stream_id, topic_name)
    "zerver_usertopic_stream_topic_user_visibility_idx" btree (stream_id, topic_name, visibility_policy, user_profile_id)
    "zerver_usertopic_user_visibility_idx" btree (user_profile_id, visibility_policy, stream_id, topic_name)
```

But since that may be a bit of a project, you can probably unblock this by just using `mock.patch` as shown in the commented-out test to make the first check inoperative, and then remove the extra testing code when we fix the index.

@abhijeetbodas2001 want to open an issue to track improve the `UserTopic` database-level constraint?

Reply:
Tim Abbott Said:
> But I think the right answer is that we should ONLY have the database-level integrity check

I think we should only have `topic_is_muted()` test -- test using Django.
We should **remove** database-level integrity check.

**Reason:**
Earlier in [set_user_topic_visibility_policy()](https://github.com/zulip/zulip/pull/24236/files#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR108), we were using:

```python
    UserTopic.objects.bulk_create(
        [
            UserTopic(
                user_profile=user_profile,
                stream_id=stream_id,
                recipient_id=recipient_id,
                topic_name=topic_name,
                last_updated=date_muted,
                visibility_policy=UserTopic.MUTED,
            ),
        ],
        ignore_conflicts=ignore_duplicate,
    )
 ```
It creates a new object or raises an Integrity Error.

Now, we are planning to use: (b/c we have now other visibility policies, other than just MUTE)
```python
        UserTopic.objects.bulk_create(
            [
                UserTopic(
                    user_profile=user_profile,
                    stream_id=stream_id,
                    recipient_id=recipient_id,
                    topic_name=topic_name,
                    last_updated=date_added,
                    visibility_policy=visibility_policy,
                ),
            ],
            ignore_conflicts=ignore_duplicate,
            update_conflicts=update_duplicate,
            update_fields=["visibility_policy"],
            unique_fields=["user_profile", "stream", "topic_name"],
        )
 ```
It creates or updates (if **{user_profile, stream, topic_name}** triplet already exists).

[Django Docs](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-create) -- related to new parameters being used.

DB will no longer raise `Integrity Error`

@timabbott your views on this ?

Reply:
[CZO Discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Follow.20Topics/near/1506164)

Reply:
This was resolved via #24354.